torchtune/modules/model_fusion/_fusion.py

```diff
@@ -463,18 +500,294 @@ def forward(
+                of each token. During training, this is used to indicate the positions
+                of each token relative to its sample when packed, shape ``[b x s]``.
+                During inference, this indicates the position of the current token.
+                If none, assume the index of the token is its position id. Default is None.
+            **kwargs (Dict[str, Any]): additional keyword arguments. This is solely used to match the
+                :class:`~torchtune.modules.TransformerDecoder` forward and does not have any effect.
+
+        Note: At the very first step of inference, when the model is provided with a prompt,
+        ``input_pos`` would contain the positions of all of the tokens in the prompt
+        (eg: ``torch.arange(prompt_length)``). This is because we will need to compute the
+        KV values for each position.
+
+        Returns:
+            torch.Tensor: output tensor with shape ``[b x s x v]`` or a list of layer \
+                output tensors defined by ``output_hidden_states`` with the \
+                final output tensor appended to the list.
+
+        Raises:
+            ValueError: if ``encoder_input`` keys do not match ``encoders`` keys
+
+        Notation used for tensor shapes:
+            - b: batch size
+            - s: token sequence length
+            - s_e: encoder sequence length
+            - v: vocab size
+            - d: token embed dim
+            - d_e: encoder embed dim
+            - m_s: max seq len
+        """
+        if encoder_input is not None and encoder_input.keys() != self.encoders.keys():
+            raise ValueError(
+                f"Found mismatched keys in encoder_input and instantiated encoders. "
+                f"Got {encoder_input.keys()}, expected {self.encoders.keys()}."
+            )
+
+        bsz, seq_len = tokens.shape
+        # is_text: [bsz, seq_len], text_embeds: [num_text, embed_dim]
+        is_text, text_embeds = self._decoder_embed(tokens)
+        embed_dim = text_embeds.shape[-1]
+
+        # Holds the final embedding vector
+        fused_embeds = torch.empty(
+            bsz, seq_len, embed_dim, dtype=text_embeds.dtype, device=text_embeds.device
+        )
+        # Place the text-only embeddings
+        fused_embeds = fused_embeds.masked_scatter(is_text.unsqueeze(-1), text_embeds)
+
+        for encoder, inp in (encoder_input or {}).items():
+            # [bsz, num_encoder_tokens, embed_dim]
+            encoder_embeds = self.encoders[encoder](**inp)
+            # [bsz * num_encoder_tokens, embed_dim]
+            encoder_embeds = encoder_embeds.view(-1, embed_dim)
+            # [bsz, seq_len, 1]
+            encoder_mask = (tokens == self.encoder_tokens[encoder]).unsqueeze(-1)
+            # At locations where encoder token is found, replace with encoder embedding
+            fused_embeds = fused_embeds.masked_scatter(encoder_mask, encoder_embeds)
```

Comment:
Sorry I might be missing the point here.. is this changing the shape of fused_embeds?

Reply:
It is not, it is just placing the encoder embedding vectors in each instance of that encoder token in the fused_embeds